File: sdk/python/aggregate.py
"""TTP trust score aggregation — the normative algorithm.

Implements protocol/aggregation-spec.md v1.1 step for step, mirroring
packages/pctr/src/aggregate.mjs. The same eleven test vectors run against both
bindings (scripts/check-agt-parity.mjs), so they cannot drift apart.

Two properties, both deliberate:
  - Negative signals weigh more (default 1.5x), so an agent behaving well most of the
    time cannot average away a few dangerous actions.
  - No single issuer may contribute more than effective_cap of the score, and the excess
    goes to the *uncapped* issuers. v1.0 re-normalized across everyone, which handed the
    excess straight back whenever the other issuers were light.
"""

from __future__ import annotations

import math
from typing import Any, Dict, List, Mapping, Optional

__all__ = ["DEFAULT_PARAMS", "INSUFFICIENT_TRUST_DATA", "aggregate_trust", "score_label"]

DEFAULT_PARAMS: Dict[str, float] = {
    "receipt_window_s": 300,
    "max_issuer_weight": 0.40,
    "negative_weight_multiplier": 1.5,
    "decay_half_life_s": 120,
}

INSUFFICIENT_TRUST_DATA = "INSUFFICIENT_TRUST_DATA"


def aggregate_trust(receipts: Optional[List[Mapping[str, Any]]] = None,
                    current_time_ms: Optional[int] = None,
                    params: Optional[Mapping[str, float]] = None) -> Dict[str, Any]:
    """Aggregate behavioural receipts into a trust score in [0.0, 1.0]."""
    p = {**DEFAULT_PARAMS, **(params or {})}
    receipts = list(receipts or [])

    # Step 1 — filter to the receipt window.
    window = [r for r in receipts
              if current_time_ms - r["timestamp"] <= p["receipt_window_s"] * 1000]
    if not window:
        return {"error": INSUFFICIENT_TRUST_DATA, "score": None,
                "contributing_receipts": 0, "contributing_issuers": 0}

    # Steps 2 and 3 — time decay, then negative signal amplification.
    weighted = []
    for r in window:
        age_s = (current_time_ms - r["timestamp"]) / 1000
        decay = math.exp(-math.log(2) * age_s / p["decay_half_life_s"])
        negative = r["score"] < 0.5
        weighted.append({**r, "age_s": age_s,
                         "signal_weight": decay * p["negative_weight_multiplier"] if negative else decay})

    # Step 4 — per-issuer weighted score.
    by_issuer: Dict[str, Dict[str, float]] = {}
    for r in weighted:
        e = by_issuer.setdefault(r["issuer_id"], {"weighted_sum": 0.0, "total_weight": 0.0})
        e["weighted_sum"] += r["score"] * r["signal_weight"]
        e["total_weight"] += r["signal_weight"]

    issuers = [{"issuer_id": k, "issuer_score": e["weighted_sum"] / e["total_weight"],
                "issuer_raw_weight": e["total_weight"]}
               for k, e in by_issuer.items()]

    # Step 5 — cap, then water-fill the excess onto the uncapped issuers. A cap below
    # 1/n is infeasible, so that is the floor on the cap actually applied.
    total_raw = sum(i["issuer_raw_weight"] for i in issuers)
    effective_cap = max(p["max_issuer_weight"], 1 / len(issuers))
    for i in issuers:
        i["weight"] = i["issuer_raw_weight"] / total_raw
        i["capped"] = False

    for _ in range(len(issuers) + 1):
        over = [i for i in issuers if not i["capped"] and i["weight"] > effective_cap + 1e-12]
        if not over:
            break
        excess = 0.0
        for i in over:
            excess += i["weight"] - effective_cap
            i["weight"] = effective_cap
            i["capped"] = True
        free = [i for i in issuers if not i["capped"]]
        free_total = sum(i["weight"] for i in free)
        if not free or free_total == 0:
            break
        for i in free:
            i["weight"] += excess * (i["weight"] / free_total)

    # Step 6 — combine, and clamp for floating point.
    raw_score = sum(i["issuer_score"] * i["weight"] for i in issuers)

    return {
        "score": max(0.0, min(1.0, raw_score)),
        "contributing_receipts": len(window),
        "contributing_issuers": len(issuers),
        "oldest_receipt_age_s": round(max(r["age_s"] for r in weighted)),
        "issuers": [{"issuer_id": i["issuer_id"],
                     "issuer_score": round(i["issuer_score"], 6),
                     "weight": round(i["weight"], 6),
                     "capped": i["capped"]} for i in issuers],
    }


# The scale from protocol/scoring-semantics.md, so a score can be read in words.
def score_label(score: Optional[float]) -> str:
    if score is None:
        return "Unknown"
    if score >= 0.90:
        return "Excellent"
    if score >= 0.70:
        return "Good"
    if score >= 0.50:
        return "Marginal"
    if score >= 0.30:
        return "Poor"
    if score >= 0.10:
        return "Bad"
    return "Critical"

File: packages/pctr/src/aggregate.mjs
// TTP TRUST SCORE AGGREGATION — the normative algorithm.
//
// Implements protocol/aggregation-spec.md v1.0 step for step. That document is marked
// normative and carries test vectors; tests/aggregation.test.mjs runs every one of them,
// so this file is not free to drift from the spec.
//
// The two properties that matter, both deliberate:
//   - Negative signals weigh more (default 1.5x). An agent behaving well most of the time
//     must not be able to average away a few dangerous actions.
//   - No single issuer may contribute more than a fraction of the score (default 0.40),
//     so one chatty or captured issuer cannot decide an agent's trust alone.

export const DEFAULT_PARAMS = {
  receipt_window_s: 300,
  max_issuer_weight: 0.40,
  negative_weight_multiplier: 1.5,
  decay_half_life_s: 120
};

export const INSUFFICIENT_TRUST_DATA = 'INSUFFICIENT_TRUST_DATA';

/**
 * @param receipts  [{ receipt_id, issuer_id, score, timestamp }] — valid, deduplicated
 * @param current_time_ms  evaluation time
 * @returns { score, contributing_receipts, contributing_issuers, oldest_receipt_age_s }
 *          or { error: 'INSUFFICIENT_TRUST_DATA' } when nothing is in the window.
 */
export function aggregateTrust(receipts = [], current_time_ms = Date.now(), params = {}) {
  const { receipt_window_s, max_issuer_weight, negative_weight_multiplier, decay_half_life_s } =
    { ...DEFAULT_PARAMS, ...params };

  // Step 1 — filter to the receipt window.
  const windowReceipts = receipts.filter(
    (r) => current_time_ms - r.timestamp <= receipt_window_s * 1000
  );
  if (!windowReceipts.length) {
    return { error: INSUFFICIENT_TRUST_DATA, score: null, contributing_receipts: 0, contributing_issuers: 0 };
  }

  // Steps 2 and 3 — time decay, then negative signal amplification.
  const weighted = windowReceipts.map((r) => {
    const age_s = (current_time_ms - r.timestamp) / 1000;
    const decay_weight = Math.exp((-Math.LN2 * age_s) / decay_half_life_s);
    const negative = r.score < 0.5;
    return {
      ...r, age_s,
      adjusted_score: r.score,
      signal_weight: negative ? decay_weight * negative_weight_multiplier : decay_weight
    };
  });

  // Step 4 — per-issuer weighted score.
  const byIssuer = new Map();
  for (const r of weighted) {
    const entry = byIssuer.get(r.issuer_id) ?? { weightedSum: 0, totalWeight: 0 };
    entry.weightedSum += r.adjusted_score * r.signal_weight;
    entry.totalWeight += r.signal_weight;
    byIssuer.set(r.issuer_id, entry);
  }
  const issuers = [...byIssuer].map(([issuer_id, e]) => ({
    issuer_id,
    issuer_score: e.weightedSum / e.totalWeight,
    issuer_raw_weight: e.totalWeight
  }));

  // Step 5 — cap each issuer's fraction, then re-normalize.
  const totalRawWeight = issuers.reduce((sum, i) => sum + i.issuer_raw_weight, 0);
  const capped = issuers.map((i) => ({
    ...i,
    capped_fraction: Math.min(i.issuer_raw_weight / totalRawWeight, max_issuer_weight)
  }));
  const normalizationFactor = capped.reduce((sum, i) => sum + i.capped_fraction, 0);

  // Step 6 — combine, and clamp for floating point.
  const rawScore = capped.reduce(
    (sum, i) => sum + i.issuer_score * (i.capped_fraction / normalizationFactor), 0
  );

  return {
    score: Math.max(0, Math.min(1, rawScore)),
    contributing_receipts: windowReceipts.length,
    contributing_issuers: issuers.length,
    oldest_receipt_age_s: Math.round(Math.max(...weighted.map((r) => r.age_s))),
    issuers: capped.map((i) => ({
      issuer_id: i.issuer_id,
      issuer_score: Number(i.issuer_score.toFixed(6)),
      weight: Number((i.capped_fraction / normalizationFactor).toFixed(6)),
      capped: i.issuer_raw_weight / totalRawWeight > max_issuer_weight
    }))
  };
}

// The scoring scale from protocol/scoring-semantics.md, so a score can be read in words.
export function scoreLabel(score) {
  if (score == null) return 'Unknown';
  if (score >= 0.90) return 'Excellent';
  if (score >= 0.70) return 'Good';
  if (score >= 0.50) return 'Marginal';
  if (score >= 0.30) return 'Poor';
  if (score >= 0.10) return 'Bad';
  return 'Critical';
}
